### ISSUE TITLE
hackport-0.6.4: status and list are broken

### ISSUE THREAD

**Issue by pbarill** [gentoo-haskell/hackport#63]
I recently updated to 0.6.4. `merge` works, `update` seems to do what it's supposed to, but I have issues with `list` and `status`.

`list`: It does the lookup as usual, but without the `*` indicating that it is present in the overlay.

`status`: I tried `hackport status --from-hackage`, and the result consists of color codes only, no package listed.

**Comment by solpeth** [comment]
I did notice this during porting, and subsequently forgot about it rather than raising an issue... So thank you!

Earliest I can have a good crack at it is the weekend.

**Comment by pbarill** [comment]
Another very (very) minor thing:
```
$ hackport -V
hackport version 0.6.4
using cabal-install 2.1.0.0 and the Cabal library version 1.9999
```
Me thinks these numbers on the last line deserve some bump!

**Comment by solpeth** [comment]
To my knowledge these are the version numbers provided when `Cabal` is bootstrapped. In reality, `0.6.4` is based on `Cabal-3`.

**Comment by solpeth** [comment]
@pbarill was this definitely working in 0.6.3?

**Comment by pbarill** [comment]
I am quite sure that it worked. Better, let me confirm:
````
emerge -1 =app-portage/hackport-0.6.3::haskell
rm -v $(portageq get_repo_path / haskell)/.hackport/hackage.haskell.org/*
hackport update
hackport list diagrams-gtk     # got a star at 1.4
````
`hackport status` works too.

This is with ghc-8.6.5 and cabal-2.4.0.1, if it makes any difference. This line above with `rm -v` is not something I thought of doing with 0.6.4, as I expected that `update` by itself should do the job.

**Comment by solpeth** [comment]
Okay, thanks. I'm going to start with trying to build  c867f17 (pre Cabal-3) and see if we can narrow down the problematic commit. If you're keen feel free to give it a go yourself, otherwise I'll get to it when I can.

**Comment by solpeth** [comment]
c867f17 didn't work for me. I'll start a git bisect.

**Comment by solpeth** [comment]
Okay. The faulty commit is likely somewhere between d36d3fb86ad72abf583c5fb3efc7a0e9b2bf0ef8 (bad) and ab160a4718b726f6c3bc56e538704ec69a3e8310 (good). The commits in between unfortunately don't build.

Bisect log:

# bad: [c867f1711bcd09cd9f73fe6ca6bb4f133389b668] Portage/GHCCore.hs: port away from Distribution.Text
# good: [a1f61b89c3c054f1dcccfbe28504d55df91487ab] hackport.cabal: release 0.6.3
git bisect start 'HEAD' 'v0.6.3'
# bad: [d36d3fb86ad72abf583c5fb3efc7a0e9b2bf0ef8] Portage/PackageId.hs: port parseFriendlyPackage to Parsec
git bisect bad d36d3fb86ad72abf583c5fb3efc7a0e9b2bf0ef8
# good: [ab160a4718b726f6c3bc56e538704ec69a3e8310] Main.hs, Status.hs: replace die with die' or dieNoVerbosity
git bisect good ab160a4718b726f6c3bc56e538704ec69a3e8310
# skip: [d778af9565744fa5ae661c5a25bcf6fa5b022b21] PackageId.hs, Version.hs: port away from Distribution.Text
git bisect skip d778af9565744fa5ae661c5a25bcf6fa5b022b21
# skip: [c66fa68558414726e779078a27b294e10b321a80] Portage/Dependency/Print.hs: port to Distribution.Pretty
git bisect skip c66fa68558414726e779078a27b294e10b321a80
# skip: [6f7fb7581f3f0efc64524fa20b196d71a60f7d57] Portage/Overlay.hs: port to Distribution.Parsec.Class
git bisect skip 6f7fb7581f3f0efc64524fa20b196d71a60f7d57
# skip: [7746d7e262ba86f0b84804d5372fe8abe5b2a628] Portage/Use.hs: port to Distribution.Pretty
git bisect skip 7746d7e262ba86f0b84804d5372fe8abe5b2a628
# good: [ab160a4718b726f6c3bc56e538704ec69a3e8310] Main.hs, Status.hs: replace die with die' or dieNoVerbosity
git bisect good ab160a4718b726f6c3bc56e538704ec69a3e8310

I'll have to dig through here to see what the problem could be.

There shouldn't be anything to do with the underlying cabal or hackage-security submodules, as they weren't touched between the good and bad commits. I'm guessing something to do with the pretty printer, at this stage.

**Comment by pbarill** [comment]
Yes I though too it's not on the cabal side, since it does read and print stuff from the hackage database, and that's great news (fixing that would be very, very cheesy). I think it's a bit deeper than pretty printer. ghci adventures:
````
z <- Portage.loadLazy "/home/pbl/overlays/gentoo-haskell/" #you have a different path
Portage.inOverlay z (PackageIdentifier (mkPackageName "hashable") (mkVersion [1,2,7,0]))
````
That returns `False`, and it shouldn't be so. Glancing at the output of `z`, it seems there's an empty list where there should be a list of versions sitting in the overlay.

**Comment by solpeth** [comment]
Nice pick up! I'm again on the way to work, the next I can have a look at this is tomorrow.

**Comment by solpeth** [comment]
I'm just thinking that this could be to do with parsing, given that parsing a package name no longer works with the `Parsec` instances provided by Cabal.

When I can eventually get to a computer I'll see if adjusting any of our `Parsec` instances for say, `PackageId` makes a difference.

**Comment by solpeth** [comment]
Got it. That seems to have been the issue. I'll push a fix shortly.

**Comment by solpeth** [comment]
@pbarill , could you please try out `app-portage/hackport-0.6.4-r1` and let me know if that revision fixes these issues? It's just `hackport-0.6.4` with 1bf8b904d9d50e49d37a951e0f672948dfa087f2 applied.

**Comment by pbarill** [comment]
I confirm that it solves the issue. And what a patch: _one line!!_
